kafka/kafka.sh

```diff
@@ -24,49 +28,78 @@ function update_apt_get() {
+    | cut -d '=' -f 2 \
+    | cut -d ':' -f 1 \
+    | sed "s/$/:${ZOOKEEPER_CLIENT_PORT}/" \
+    | xargs echo  \
+    | sed "s/ /,/g")
+
+  if [[ -z "${ZOOKEEPER_LIST}" ]]; then
+    # Didn't find zookeeper quorum in zoo.cfg, but possibly workers just didn't
+    # bother to populate it. Check if YARN HA is configured.
+    ZOOKEEPER_LIST=$(bdconfig get_property_value --configuration_file \
       /etc/hadoop/conf/yarn-site.xml \
       --name yarn.resourcemanager.zk-address 2>/dev/null)
-fi
-
-# If all attempts failed, error out.
-if [[ -z "${ZOOKEEPER_LIST}" ]]; then
-  echo 'Failed to find configured Zookeeper list; try --num-masters=3 for HA'
-  exit 1
-fi
-
-# Install Kafka from Dataproc distro.
-apt-get install -y kafka-server || dpkg -l kafka-server
-
-mkdir -p /var/lib/kafka-logs
-chown kafka:kafka -R /var/lib/kafka-logs
-
-# Note: If modified to also run brokers on master nodes, this logic for
-# generating BROKER_ID will need to be changed.
-BROKER_ID=$(hostname | sed 's/.*-w-\([0-9]\)*.*/\1/g')
-sed -i 's|log.dirs=/tmp/kafka-logs|log.dirs=/var/lib/kafka-logs|' /etc/kafka/conf/server.properties
-sed -i 's|^\(zookeeper\.connect=\).*|\1'${ZOOKEEPER_LIST}'|' /etc/kafka/conf/server.properties
-sed -i 's,^\(broker\.id=\).*,\1'${BROKER_ID}',' /etc/kafka/conf/server.properties
-echo 'delete.topic.enable = true' >> /etc/kafka/conf/server.properties
-
-# Start Kafka
-service kafka-server restart
+  fi
+
+  # If all attempts failed, error out.
+  if [[ -z "${ZOOKEEPER_LIST}" ]]; then
+    err 'Failed to find configured Zookeeper list; try --num-masters=3 for HA'
+    exit 1
+  fi
+
+  # Install Kafka from Dataproc distro.
+  apt-get install -y kafka-server \
+    || (err 'Unable to install kafka-server' && exit 1)
+
+  mkdir -p /var/lib/kafka-logs
+  chown kafka:kafka -R /var/lib/kafka-logs
+
+  # Note: If modified to also run brokers on master nodes, this logic for
+  # generating BROKER_ID will need to be changed.
+  BROKER_ID=$(hostname | sed 's/.*-w-\([0-9]\)*.*/\1/g')
+  sed -i 's|log.dirs=/tmp/kafka-logs|log.dirs=/var/lib/kafka-logs|' \
+    "${KAFKA_PROP_FILE}"
+  sed -i 's|^\(zookeeper\.connect=\).*|\1'${ZOOKEEPER_LIST}'|' \
+    "${KAFKA_PROP_FILE}"
+  sed -i 's,^\(broker\.id=\).*,\1'${BROKER_ID}',' \
+    "${KAFKA_PROP_FILE}"
+  echo 'delete.topic.enable = true' >> "${KAFKA_PROP_FILE}"
+
+  # Start Kafka.
+  service kafka-server restart
+}
+
+function main() {
+  local ROLE="$(/usr/share/google/get_metadata_value attributes/dataproc-role)"
+  update_apt_get || (err 'Unable to update packages lists.' && exit 1)
+
+  # Only run the installation on workers; verify zookeeper on master(s).
+  if [[ "${ROLE}" == 'Master' ]]; then
+    service zookeeper-server status \
+      || (err 'Required zookeeper-server not running on master!' && exit 1)
+    # On master nodes, just install kafka libs but not kafka-server.
+    apt-get install -y kafka \
+      || (err 'Unable to install kafka libraries on master!' && exit 1)
+    exit 0
```

Comment:
Please use `if`/`else` rather than using `exit` for control flow.

Reply:
Sure.